### PR TITLE
Share mouse position of players with observers

### PR DESCRIPTION
**Pull request by Garanas**
A useful feature for casters: being able to see where the mouse of the players is pointing at! The feature isn't smooth yet, still needs some polishing. But the idea works and it is cheat free. We 'abuse' the chat system that allows you to send any data to any client. The data is properly culled by the engine, as a result players that should not receive the data has no idea that information has been send. There's no way to retrieve this via the UI.

https://user-images.githubusercontent.com/15778155/187092955-a22027f5-af25-44fe-868a-f1e801b6db97.mp4

Things to take care of:

- [ ] Use army color for the mesh
- [x] Take into account the 500 ms delay, this should also help with the lag issues

And nice to have features:

- [ ] Find out if we can skip an update if the mouse didn't move

This pull requests adds in the tracking of the mouse of players. But similarly, we could track their camera settings (to see what they are seeing) and other interesting information that can be used for casting.

### CODE REVIEW

**Comment by Garanas** [comment]
https://user-images.githubusercontent.com/15778155/187370290-37c3917b-716e-408e-a476-e5dff3e16a9b.mp4

This version cheats a little, but it works like a charm! Still need to refactor it, once that is done then we're good to go for a live trial!

File: lua/UserSync.lua [hunk 1]
```diff
@@ -1,3 +1,5 @@
+---@meta
```
[the remark this answers is not in the document]
Comment (Garanas) [reply lua/UserSync.lua:1]: this file is globally available though

[the remark this answers is not in the document]
Comment (Garanas) [reply lua/UserSync.lua:1]: Should I otherwise just mark it as @global?

[the remark this answers is not in the document]
Comment (Garanas) [reply lua/UserSync.lua:1]: woops, I apparently pinged someone doing that 😄 , apologies

**Comment by Garanas** [comment]
Merging this for live testing.

**Comment by Garanas** [comment]
Done some live testing, it works like a charm!

It appears however that this will never work when running via a replay. The send via this approach is not stored in the replay. The reason we can see chat messages (that use the same approach is because:

- Once received, it does a sim callback: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/ui/game/chat.lua#L810
- Which leads to: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/SimCallbacks.lua#L448
- Which then sends it back up to the UI: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/SimUtils.lua#L656
- See also: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/SimUtils.lua#L651

And that actually shows the message. That is why chat messages are stored in the replay: they are sim callbacks! It is also the reason why what we're doing here will likely never be in the replay. We can't do a sim callback each tick. The current approach would only work when you are observing the game directly.

**Comment by Hdt80bro** [comment]
Well, we still *could* use a sim callback to store the mouse position for every frame in the sim tick to the replay and just not send it back to the UI--it'd involve separate code to receive the data though and that sounds like a lot more work

**Comment by Hdt80bro** [comment]
Yes, we would need to do some crazy optimizations and math approximations to get it down to size. It *could* be done viably, but not this year.

**Comment by Garanas** [comment]
Yeah, lets stick to what we have now. During the test it worked like a charm, it is very exciting 👍

**Comment by Garanas** [comment]
I'm going to do a few more PRs to make it look better (match player color, etc) and to add some form of camera tracking too.

**Comment by Garanas** [comment]
What I was thinking of to make the feature better:

- [ ] Add player-color meshes / labels
- [ ] Add a 'blink' when a player issues a command (via `OnCommandIssued` of `commandmode.lua`), we can have different colors for different type of commands
- [ ] Add a 'trace' of meshes that trace behind the cursor, to show how quickly it moves
- [ ] Add a 'disabled' version of the mesh when the mouse is on top of UI (it doesn't update the world position when it is)

And of course, being able to adjust how it looks / works

**Comment by Garanas** [comment]
https://user-images.githubusercontent.com/15778155/187546231-1e2f0ea9-9dc0-4f16-90f0-0dc6ee9f5377.mp4

https://user-images.githubusercontent.com/15778155/187546262-126ce99d-5e70-4321-b8ec-878ffb8f0d13.mp4

https://user-images.githubusercontent.com/15778155/187546268-e21fb4be-4783-4939-9b15-28209cdf3021.mp4

https://user-images.githubusercontent.com/15778155/187546278-79b3bbf1-d71d-49e7-9397-bf430e225786.mp4

**Pull request merged**